app/controllers/saved_pages_controller.rb

```diff
@@ -0,0 +1,72 @@
+class SavedPagesController < ApplicationController
+  include AuthenticatedApiConcern
+
+  before_action :check_saved_page_exists, only: %i[destroy show]
+
+  # GET /api/saved_pages
+  def index
+    @saved_pages = SavedPage.where(oidc_user_id: user.id)
+    render_api_response(saved_pages: present_pages(@saved_pages))
+  end
+
+  # GET /api/saved_pages/:page_path
+  def show
+    render_api_response(saved_page: present_page(saved_page))
+  end
+
+  # POST /api/saved_pages
+  def create
+    saved_page = SavedPage.new(oidc_user_id: user.id, page_path: params[:page_path])
+
+    if saved_page.save
+      render_api_response(saved_page: present_page(saved_page))
+    else
+      handle_create_saved_page_errors(saved_page)
+    end
+  end
+
+  # DELETE /api/saved_pages
+  def destroy
+    return head :no_content if saved_page.destroy!
+
+    raise ApiError::SavedPageNotDestroyed, { page_path: params[:page_path] }
+  end
+
+private
+
+  def handle_create_saved_page_errors(saved_page)
+    error_types = saved_page.errors.map(&:type)
+
+    if error_types.include?(:taken)
+      raise ApiError::SavedPageDuplicate, { page_path: params[:page_path] }
+    end
```

Comment:
I'm in two minds about this.  On the one hand, saving the same page twice is a bit weird and probably shouldn't happen; but on the other hand is there any downside to making `create` do an insert-or-update?

We're not doing something like assigning each saved page a unique ID: the path is the ID.  So the API state after one POST is the same as it is after 10 POSTs.  The `create` method is idempotent.

Reply:
Yeah I see what you mean, Initially i'd wanted this for the alert we could send the user, essentially:
> your trying to save something that is already saved, don't worry it's we've got the thing you want on file, here it is.

just saving it and showing the success state achieves the same thing but cleaner...
🤔 but I'm wondering about the code we'll inevitably need to add for what happens if guidance is `gone` or `[withdrawn]` or redirected.

Updates there might mean they go to create something then find out its a duplicate later?
Wouldn't a specific message that highlights that be useful there? If so, this pattern solves a fairly rare problem now and might make that easier later?

Reply:
> your trying to save something that is already saved, don't worry it's we've got the thing you want on file, here it is.

Is that useful and actionable information for the user?  What will someone do differently in the case that a page is already saved?

> Updates there might mean they go to create something then find out its a duplicate later?

If the only route to save a page is via a button on the page which shows a save/unsave state based on the `show` response, then the only way a user could save a page twice is if they had two tabs with the same page open, or if there were a bug in the button logic.

If, for example, a page gets redirected and we update the path in the database automatically, then when the user looks at the new page the button will be in the saved state.

Reply:
Ok so then it's back to a POST updating things 🤔 
So to run down this, to impliment this I think i'd be looking at:
- removing the unique index on the table
- removing the duplicate validation
- changing the post controller to use find_or_create_by
- removing the error, and the docs

If we're on the fence about it, then better to leave it?
If we've swung towards removing it in this thread then I'll go remove those parts?

Reply:
> removing the unique index on the table
> removing the duplicate validation

We definitely still want the uniqueness, to protect against application bugs.  But I don't think we want to treat it as an API error to save the same page twice.